### DISCUSSION TITLE
Compiling on modern linux versions

### DISCUSSION THREAD

**Discussion by phirestalker** [theophanemayaud/video-simili-duplicate-cleaner#113] [Q&A]
Great program. I bought it for my mac machine, but I also wish to use it on my linux machine. Debian bookworm on arm64. The problem is that it is using newer libraries and they are apparently not compatible. I was able to get rid of the opencv2 include files not being found by adding to the include path. However, now I seem to be getting a lot of errors related to ffmpeg.

Has anyone got this to compile on a modern linux?
Is there a binary somewhere for linux arm?
Is there any hope of some library dependency updates to this program?

**Comment by theophanemayaud** [comment]
What ffmpeg libs are you linking to when compiling the app ? I've never tried for linux, if you get it to work please submit a PR it'd be great as an example for others !

I believe it should be pretty similar to macOS : 
- compile ffmpeg for your os/architecture
- compile the opencv2 library for your boss/architecture
- compile the main app

I really had trouble compiling ffmpeg on windows but on macOS it was ok, and linux should be even better. opencv2 if I remember correctly I found the libs precompiled. 
Then it's just a matter of putting in qt the correct paths in the correct places...

**Reply by theophanemayaud** [reply]
I put all the info I had when first doing it here : https://github.com/theophanemayaud/video-simili-duplicate-cleaner/blob/master/DEPENDENCIES.md

**Reply by phirestalker** [reply]
Debian has version 5 installed. I will follow the instructions you linked here and see if I can get it compiled.

**Reply by phirestalker** [reply]
How do I get it to use the compiled ffmpeg version instead of the installed one? I added it to LIBS in the pri file, but since it is added at the end (+=) it is still trying to use the installed version.

**Reply by theophanemayaud** [reply]
I’m not sure, i don’t remember more than what I wrote two years ago on that memo. 
It’s probably explained somewhere : how to link a library with qt and qmake as the project is not yet using cmake. Please link any helpful info you can find for later reference ⭐️

**Reply by EldrinSMP** [reply]
How did compiling this for Linux end up? I'm moving to a Linux system and this is an invaluable tool.

**Reply by theophanemayaud** [reply]
Hey @EldrinSMP i never compiled for Linux but you can start by installing QT, and looking at the dependencies.md to get inspiration on compiling the ffmpeg and opencv dependencies. If you have any specific questions don’t hesitate to ask 😊
I hope @phirestalker answers as well !

**Reply by phirestalker** [reply]
I'm afraid I gave up. I don't know enough to update the code to utilize a newer ffmpeg, or how to get it to use a specific version I download. Just adding to the INCLUDEPATH etc. will not work.

**Reply by theophanemayaud** [reply]
You need to read and adapt for linux the DEPENDENCIES.md information, in order to build or download native ffmpeg and opencv libraries. This is the most annoying part, why I also didn't try to support linux, because the main code is much easier to build with qt for each platform, but the dependencies are such a pain !

**Reply by EldrinSMP** [reply]
@theophanemayaud Oh, is THAT all.  LOL.  Is it bad that even though I've never coded anything more than config files and the like, and really don't have THAT much compiling experience, and COMPLETELY failed when I tried to compile the Windows app... I kinda want to try? LOL I actually compiled a custom version of FFmpeg and FFprobe this morning to try to get my OTHER video duplicate finder to work on Linux, and the process really was painless.  With the various make tools and pretty much everything in my distro being QT5 (soon to be 6) based, who knows, I might get lucky.  I'd have to have access to how you set the QT stuff up to compile it, though. That's where i lost the plot with the Windows version. I couldn't find your records on here.

 Since I switched over to Solus Linuix, I have learned more about the inner workings of an operating system and how programs are actually built than I have in the 30 years I've used DOS then Windows.  So... uhh, is there a bounty on this?  :-D

**Reply by theophanemayaud** [reply]
Yeah compilation is quite the pain ! If you have any specific questions I’m happy to help.

**Reply by taw123** [reply]
@EldrinSMP Any chance of turning your build into a Container? Would be a lifesaver for those of us trying to run it on a NAS or other Linux system in Docker.   

I've been trying to get this (simili), Video Duplicate Finder, or even the old/abandoned vidup to work to no avail. I am/was actually a s/w Engineer for years (at a slightly known fruit company) so I am not COMPLETELY without skills though it has been years...

Having also worked in Evangelism for years I have no qualms about happily PURCHASING Simili but unfortunately you don't have a version that runs natively on Linux (thus my comment about creating a container). I am looking to use this (or something that actually supports QSV hardwire decode) to search for similar video content natively on my NAS with an option to generate/rebuild hard links on demand to deDup. Would be nice if it also supported non-video (like perhaps DupGuru or even Czkawka (I am actually trying Czkawka now for non-video files as it's video support are a bit limited for my use).

Any hints would be helpful.

Thanks for all the efforts!
--T

**Reply by EldrinSMP** [reply]
@taw123 I unfortunately haven't had any time to work on this, as you can see I'm just now getting back to it. I actually purchased this program for Windows... and then a month later jumped ship to Linux and haven't looked back. LOL  If you need a Linux solution you might want to try videoduplicatefinder. It's cross platform, in active development, and does a decent job of identifying duplicates. On Windows I ran it alongside Simili and between the two that caught almost everything.

**Reply by EldrinSMP** [reply]
Well, it's been over a year, but I am proud to announce that I have made some strides in compiling the app for Linux... The only issue is it doesn't work. Everything compiles and I am able to launch the program, it'll scan all of my files, but the moment ffmpeg kicks in to be gin the comparisons it looses the plot. Throwing errors like crazy and crashing, so I am ASSUMING that it has something to do with my ffmpeg build. I followed the DEPENDENCIES.md instructions as closely as I could, unfortunately those instructions are a bit outdated. Instructions of how FFMPEG is compiled for the CURRENT version of the program would be fantastic, OpenCV as well.  I'm going to keep working on it.

**Reply by theophanemayaud** [reply]
Nice !!! Well, almost 😅
In the past two weeks I have been trying to get it rebuild on windows, and started using vcpkg to handle all the dependencies building. You might benefit from also trying it this way ?

look at branch cursor/setup-windows-cmake-with-vcpkg-dependencies-752a where using cmake + vcpkg, you’d have to adapt the «triplet» of the build config you want of course (possibly don’t specify any and let vcpkg build it for you with your machine’s default «triplet» I.e. machine architecture). 
To do this, just install vcpkg which on windows meant cloning the repo and running a setup script, then create a cmake preset like I have 2 for windows, and then call cmake with that preset ! Since I setup the cmake preset to use vcpkg as the toolchain file, it should automatically start by building/installing all dependencies properly !
If you have any questions please ask !

**Reply by theophanemayaud** [reply]
Just note that on my machine it takes hours and a few tens of gigabytes for vcpkg to compile all the dependencies because qt, ffmpeg and opencv somehow depend on lots of stuff ! But just run it once’s and it should be good from then !

**Reply by EldrinSMP** [reply]
It's WORKING!  

First and biggest thing I had to figure out, vcpkg, on Arch Linux at least, is a NO GO.  The versions of ffmpeg and opencv available on there just won't build and I'm still not sure why. Best I can figure out is there's some fundamental incompatibility between my personal setup and how it builds its libraries. I ended up, as a baseline, just using my system installations of all of the dependencies to compile and test everything, and using just CMAKE and ninja through QT Creator for the app itself, and it worked.

What was really needed was simple tweaking of the source code and getting the setup of the QT Project just right so that everything worked. Most of the errors I encountered were from platform specific code, which I have removed, changes in syntax between the base code creation and the newer version compilers that I was using, as well as quirks of the system installed dependencies I am using. 

I am not ready to post everything just yet, as I still need to clean everything up and I want to build custom ffmpeg and opencv libraries based on your specifications from the DEPENDENCIES.md and recompile the app using them in order to streamline everything. Plus, honestly, my repository locally is a mess and I need to get it in order as well. I'll be in contact with @theophanemayaud when I have everything cleaned up about releasing it to the community for testing.  Sorry for the huge post, but I'm really, really excited that I actually have it working. LOL

**Reply by theophanemayaud** [reply]
Nice !!!! Too bad the vcpkg route didn't work, but at least you found a way ! I'll be happy to work with you on merging the required changes as soon as you have them ~ready 👌 Thank you !!!

**Reply by EldrinSMP** [reply]
@theophanemayaud I've emailed you regarding merging, etc. and how you want to handle it.

**Reply by EldrinSMP** [reply]
Well, something broke. LOL The Linux version is no longer working. I synced with the new changes in the main repository to start over with newer code, and I noticed that the ssim.h file is missing when I went to start work. Is this a known issue?

**Reply by theophanemayaud** [reply]
Ah, well maybe it’s because of the latest work I’ve been doing on Mac and haven’t yet released as a new version of, thus haven’t yet tested the windows compilation options to check they still work. 
though… I didn’t modify ssim I think, only the tests… but I did modify how the tests depend on ssim. Could you check if it’s missing somewhere in a cmakelists for your Linux fork ?
And could you show me your fork so I can integrate it in the main repository and make it available to everyone (and also avoid breaking it unknowingly in the future 😅)

**Reply by theophanemayaud** [reply]
Hey @EldrinSMP, I just released version 1.13.1 on both macOS and windows so both compile successfully now. If you want me to take a look if you're still having trouble with building for linux, don't hesitate to ask again here :)
